core/src/main/java/org/infinispan/container/entries/MarshalledValueEntrySizeCalculator.java

```diff
@@ -0,0 +1,45 @@
+package org.infinispan.container.entries;
+
+import org.infinispan.io.ExpandableMarshalledValueByteStream;
+import org.infinispan.io.ImmutableMarshalledValueByteStream;
+import org.infinispan.io.MarshalledValueByteStream;
+import org.infinispan.marshall.core.MarshalledValue;
+import sun.misc.Unsafe;
+
+/**
+ * Entry Size calculator that returns an approximation of how much memory a marshalled value contains for a key
+ * and value.  Note that this also handles any types that MarshalledValue doesn't as well.
+ * @author wburns
+ * @since 8.0
+ */
+public class MarshalledValueEntrySizeCalculator extends PrimitiveEntrySizeCalculator {
+   @Override
+   protected long handleObject(Object object) {
+      Class<?> objClass = object.getClass();
+      if (objClass == MarshalledValue.class) {
+         MarshalledValue marshalledValue = (MarshalledValue)object;
+         // Marshalled value is an object and reference to class
+         long size = OBJECT_SIZE + POINTER_SIZE;
+         // Marshalled value has MarshalledValueByteStream & StreamingMarshaller instance fields and 2 ints
+         size = roundUpToNearest8(size + POINTER_SIZE * 2 + 4 * 2);
+         // We ignored the StreamingMarshaller instance as it is shared
+         MarshalledValueByteStream mvbs = marshalledValue.getRaw();
+         long mvbsSize;
+         if (mvbs instanceof ImmutableMarshalledValueByteStream) {
+            mvbsSize = OBJECT_SIZE + POINTER_SIZE;
+            mvbsSize = roundUpToNearest8(mvbsSize + POINTER_SIZE);
```

Comment:
Yeah, I suspected that. One has to be really knowledgeable about JVM memory model in order to understand this code. Maybe it would be useful to spell out some of these details as code comments? I suppose maintainers and people reviewing code would spot potential issues more easily. Great stuff @wburns .